claat/parser/md/parse.go

```diff
@@ -542,11 +392,432 @@ func addMetadataToCodelab(m map[string]string, c *types.Codelab) {
+	}
+	return row
+}
+
+// survey expects a title followed by 1 or more lists. They all are in the same dd element.
+func survey(ds *docState) types.Node {
+	var gg []*types.SurveyGroup
+	hn := ds.cur
+	for hn = hn.NextSibling; hn != nil; hn = hn.NextSibling {
+		ds.cur = ds.cur.NextSibling
+		if hn.DataAtom != atom.Dd {
+			continue
+		}
+		optionsNode := findAtom(hn, atom.Li)
+		if optionsNode == nil {
+			fmt.Println("No survey results list")
+			continue
+		}
+		opt := surveyOpt(optionsNode)
+		if len(opt) > 0 {
+			gg = append(gg, &types.SurveyGroup{
+				Name:    strings.TrimSpace(hn.FirstChild.Data),
+				Options: opt,
+			})
+		}
+	}
+	if len(gg) == 0 {
+		return nil
+	}
+	ds.survey++
+	id := fmt.Sprintf("%s-%d", ds.clab.ID, ds.survey)
+	return types.NewSurveyNode(id, gg...)
+}
+
+func surveyOpt(hn *html.Node) []string {
+	var opt []string
+	for ; hn != nil; hn = hn.NextSibling {
+		if hn.DataAtom != atom.Li {
+			continue
+		}
+		opt = append(opt, stringifyNode(hn, true))
+	}
+	return opt
+}
+
+// code parses hn as inline or block codes.
+// Inline code node will be of type NodeText.
+func code(ds *docState, term bool) types.Node {
+	elem := findParent(ds.cur, atom.Pre)
+	// inline <code> text
+	// TODO, here
+	if elem == nil {
+		return text(ds)
+	}
+	// block code or terminal
+	v := stringifyNode(ds.cur, false)
+	if v == "" {
+		if countDirect(ds.cur.Parent) > 1 {
+			return nil
+		}
+		v = "\n"
+	} else if ds.cur.Parent.FirstChild == ds.cur && ds.cur.Parent.DataAtom != atom.Span {
+		v = "\n" + v
+	}
+	n := types.NewCodeNode(v, term)
+	n.MutateBlock(elem)
+	return n
+}
+
+// list parses <ul> and <ol> lists.
+// It returns nil if the list has no items.
+func list(ds *docState) types.Node {
+	typ := nodeAttr(ds.cur, "type")
+	if ds.cur.DataAtom == atom.Ol && typ == "" {
+		typ = "1"
+	}
+	start, _ := strconv.Atoi(nodeAttr(ds.cur, "start"))
+	list := types.NewItemsListNode(typ, start)
+	for hn := findAtom(ds.cur, atom.Li); hn != nil; hn = hn.NextSibling {
+		if hn.DataAtom != atom.Li {
+			continue
+		}
+		ds.push(hn)
+		nn := parseSubtree(ds)
+		nn = compactNodes(nn)
+		ds.pop()
+		if len(nn) > 0 {
+			list.NewItem(nn...)
+		}
+	}
+	if len(list.Items) == 0 {
+		return nil
+	}
+	if ds.lastNode != nil {
+		switch ds.lastNode.Type() {
+		case types.NodeHeaderCheck:
+			list.MutateType(types.NodeItemsCheck)
+		case types.NodeHeaderFAQ:
+			list.MutateType(types.NodeItemsFAQ)
+		}
+	}
+	return list
+}
+
+// image creates a new ImageNode out of hn, parsing its src attribute.
+// It returns nil if src is empty.
+// It may also return a YouTubeNode if alt property contains specific substring.
+func image(ds *docState) types.Node {
+	if strings.Contains(nodeAttr(ds.cur, "alt"), "youtube.com/watch") {
+		return youtube(ds)
+	}
+	s := nodeAttr(ds.cur, "src")
+	if s == "" {
+		return nil
+	}
+
+	n := types.NewImageNode(s)
+
+	alt := nodeAttr(ds.cur, "alt")
+	if alt != "" {
+		n.Alt = alt
+	}
+
+	title := nodeAttr(ds.cur, "title")
+	if title != "" {
```

Comment:
nit: you might as well make this more compact:

```go
if v := nodeAttr(ds.cur, "title"); v != "" {
  n.Title = v
}
```

the same for `alt` above.
then, the whole image function will get quite smaller.

Reply:
Correct, we don't reuse v later on, to if scope is enough. I'm even unsure why I didn't modify it. Done now!